scripts/fit.py

```diff
@@ -0,0 +1,564 @@
+                process_axis=fitter.indata.axis_procs,
+                prefit=prefit,
+            )
+
+    if args.computeVariations:
+        if prefit:
+            cov_prefit = fitter.cov.numpy()
+            fitter.cov.assign(fitter.prefit_covariance(unconstrained_err=1.0))
+
+        exp, aux = fitter.expected_events(
+            inclusive=True,
+            compute_variance=False,
+            compute_variations=True,
+            profile_grad=False,
+        )
+
+        ws.add_expected_hists(
+            fitter.indata.channel_info,
+            exp,
+            var=aux[0],
+            variations=True,
+            prefit=prefit,
+        )
+
+        for p in args.project:
+            channel = p[0]
+            axes = p[1:]
+
+            exp, aux = fitter.expected_events_projection(
+                channel=channel,
+                axes=axes,
+                inclusive=True,
+                compute_variance=False,
+                compute_variations=True,
+                profile_grad=False,
+            )
+
+            channel_axes = fitter.indata.channel_info[channel]["axes"]
+
+            ws.add_expected_projection_hists(
+                channel,
+                axes,
+                channel_axes,
+                exp,
+                var=aux[0],
+                variations=True,
+                prefit=prefit,
+            )
+
+        if prefit:
+            fitter.cov.assign(tf.constant(cov_prefit))
+
+
+def fit(args, fitter, ws, dofit=True):
+
+    if args.externalPostfit is not None:
+        # load results from external fit and set postfit value and covariance elements for common parameters
+        with h5py.File(args.externalPostfit, "r") as fext:
+            if "x" in fext.keys():
+                # fitresult from combinetf
+                x_ext = fext["x"][...]
+                parms_ext = fext["parms"][...].astype(str)
+                cov_ext = fext["cov"][...]
+            else:
+                # fitresult from combinetf2
+                h5results_ext = io_tools.get_fitresult(fext)
+                h_parms_ext = h5results_ext["parms"].get()
+
+                x_ext = h_parms_ext.values()
+                parms_ext = np.array(h_parms_ext.axes["parms"])
+                cov_ext = h5results_ext["cov"].get().values()
+
+        xvals = fitter.x.numpy()
+        covval = fitter.cov.numpy()
+        parms = fitter.parms.astype(str)
+
+        # Find common elements with their matching indices
+        common_elements, idxs, idxs_ext = np.intersect1d(
+            parms, parms_ext, assume_unique=True, return_indices=True
+        )
+        xvals[idxs] = x_ext[idxs_ext]
+        covval[np.ix_(idxs, idxs)] = cov_ext[np.ix_(idxs_ext, idxs_ext)]
+
+        fitter.x.assign(xvals)
+        fitter.cov.assign(tf.constant(covval))
+    else:
+        fitter.profile = True
+
+        if dofit:
+            fitter.minimize()
+
+        val, grad, hess = fitter.loss_val_grad_hess()
+        fitter.hess = hess
```

Comment:
What's the motivation for having different treatment/logic of hess and cov here? (assign by python reference vs assign by tensorflow value)

In general for things which are reused within a function, using tensor.assign should be better for avoiding recompilation of the function  (but I'm not sure which is better/worse for memory usage in case of large matrices)

Reply:
Removed the hess attribute from the class